### ISSUE TITLE
Lazy is a guarantee for slowness

### ISSUE THREAD

**Issue by jrieken** [microsoft/vscode#85275]
Highly recommended read: https://philipwalton.com/articles/idle-until-urgent/

tldr; The `Lazy` util helps with speeding up creating composite types because it defers creation of properties, like [here](https://github.com/microsoft/vscode/blob/4b6e4ee176acba945da27405b41712999f46f4c9/src/vs/editor/contrib/codeAction/codeActionCommands.ts#L98). However, because it runs its executor on first access it guarantees to make that first access slow. The better alternative is the [idleCallback](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback)-api that browsers provide. We have that implemented as [`IdleValue`](https://github.com/microsoft/vscode/blob/aa7721f4381ee41d45d97dc17d9721660b5e7366/src/vs/base/common/async.ts#L714). It is equally lazy and only in the worst case as blocking as Lazy is.

**Comment by mjbvz** [comment]
Good read.

There are two places where I opted to use lazy:

- Webviews. We could probably use `IdleValue` instead but I need to confirm this. Webviews are expensive in terms of memory usage so it is best to defer creating them until they really will show to the user. However I think the lazy values we use for webview usually end up being resolved pretty quickly after being created, so `Lazy` may not be providing much benefit

-  In the code action UI. Here I use `Lazy` in order to defer creating dom nodes and hooking up event listeners until we really do have a need for them. These operations are not very expensive and I could swap in IdleValue, but again I thought it better to defer this work

**Comment by jrieken** [comment]
I'd say that IdleValue is the clear winner as soon as there is a chance of an object being instantiated, like the light bulb widget. If the code action controller knows that the language supports code actions (has a provider) it's better to use the idle value.

**Comment by mjbvz** [comment]
Closing this as there are cases—like webviews—where we really do not want to create the value ahead of time because they are quite costly. Lazy is also useful for cases where we only sometimes need to use the result of a computation
